### PR TITLE
APPROVED [MOBILE-637] Xamarin: Update Android binding to latest SDK 10

### PR DESCRIPTION
**Pull request by BrianBatchelder**
### What do these changes do?
Update Xamarin bindings and sample app for Android SDK 10

### Why are these changes necessary?
We want to provide support for the latest native SDKs.

### How did you verify these changes?
Local testing of push, message center, and IAA.

#### Verification Screenshots:
Video: https://urbanairship.slack.com/archives/G0B2N1MGX/p1560374158064900

### Anything else a reviewer should know?
I added a bunch of review comments that hopefully will help to make understanding which changes were made because of API changes, which are for the new Settings, which are for the new Preference package, etc.

FYI - the final commit (b853804) fixes formatting in some of the C# classes. It might be easier to leave it out when you review.

### CODE REVIEW

File: samples/android/HomeFragment.cs [hunk 1]
```diff
@@ -76,7 +76,7 @@ public override void OnResume()
 
 			// Use local broadcast manager to receive registration events to update the channel
 			IntentFilter channelUpdateFilter = new IntentFilter();
-			channelUpdateFilter.AddAction(UrbanAirshipReceiver.ACTION_CHANNEL_UPDATED);
+			channelUpdateFilter.AddAction(UrbanAirshipListener.ACTION_CHANNEL_UPDATED);
```
Comment (BrianBatchelder) [inline samples/android/HomeFragment.cs]: SDK 10 API change.

File: samples/android/Preference/AddTagsPreference.cs [hunk 2]
```diff
@@ -7,8 +7,11 @@
 using System.Collections.Generic;
 using System.Text;
 
+using Android.App;
```
Comment (BrianBatchelder) [inline samples/android/Preference/AddTagsPreference.cs]: The code below here looks like a complete replacement, but it is actually very similar. Certain methods moved from `DialogPreference` to `PreferenceDialogFragmentCompat`, which seems to have confused the git diff engine.

File: samples/android/Preference/NamedUserPreference.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
+﻿using System;
```
Comment (BrianBatchelder) [inline samples/android/Preference/NamedUserPreference.cs]: Our SDK doesn't expose a `NamedUserPreference`, so I created this one. If we ever do add one to the SDK, we can remove this class.

File: samples/android/Resources/xml/preferences.axml [hunk 4]
```diff
@@ -1,93 +1,56 @@
 <?xml version="1.0" encoding="utf-8"?>
-<PreferenceScreen xmlns:android="http://schemas.android.com/apk/res/android" >
```
Comment (BrianBatchelder) [inline samples/android/Resources/xml/preferences.axml]: Again, these changes look bigger than they actually are. The structure is the same, just adapted to the v7 preference syntax.

File: samples/android/Sample.csproj [hunk 5]
```diff
@@ -48,19 +48,13 @@
     <Reference Include="System.Core" />
     <Reference Include="Mono.Android" />
     <Reference Include="Xamarin.GooglePlayServices.Basement">
-      <HintPath>packages\Xamarin.GooglePlayServices.Basement.60.1142.1\lib\MonoAndroid80\Xamarin.GooglePlayServices.Basement.dll</HintPath>
+      <HintPath>packages\Xamarin.GooglePlayServices.Basement.71.1610.0-preview6\lib\monoandroid90\Xamarin.GooglePlayServices.Basement.dll</HintPath>
```
Comment (BrianBatchelder) [inline samples/android/Sample.csproj]: The FCM package requires preview versions of GooglePlayServices and Firebase to work. We may want to delay our release until the final versions are released.

File: samples/android/SampleAutopilot.cs [hunk 6]
```diff
@@ -27,6 +27,11 @@ public override void OnAirshipReady(UAirship airship)
 				sharedPreferences.Edit().PutBoolean(FirstRunKey, false).Apply();
 				airship.PushManager.UserNotificationsEnabled = true;
 			}
+
+            UrbanAirshipListener airshipListener = new UrbanAirshipListener();
+            airship.PushManager.SetNotificationListener(airshipListener);
+            airship.PushManager.AddPushListener(airshipListener);
+            airship.PushManager.AddRegistrationListener(airshipListener);
```
Comment (BrianBatchelder) [inline samples/android/SampleAutopilot.cs]: SDK 10 API changes.

File: samples/android/UrbanAirshipListener.cs [hunk 7]
```diff
@@ -15,44 +15,44 @@ namespace Sample
 	[BroadcastReceiver(Exported = false)]
 	[IntentFilter (new string[]{"com.urbanairship.push.CHANNEL_UPDATED", "com.urbanairship.push.OPENED", "com.urbanairship.push.DISMISSED", "com.urbanairship.push.RECEIVED"}, 
 		Categories = new string[]{"@PACKAGE_NAME@"})]
-	public class UrbanAirshipReceiver : AirshipReceiver
-	{
+	public class UrbanAirshipListener : Java.Lang.Object, IRegistrationListener, IPushListener, INotificationListener
```
Comment (BrianBatchelder) [inline samples/android/UrbanAirshipListener.cs]: SDK 10 API changes.

File: src/AirshipBindings.Android.Core/Additions/ActionRunRequest.cs [hunk 8]
```diff
@@ -16,7 +16,7 @@ public virtual void Run (Action<ActionArguments, ActionResult> callback)
 
 		public virtual void Run (Action<ActionArguments, ActionResult> callback, Looper looper)
 		{
-			Run (new ActionCompletionCallback (callback), looper);
+			Run (looper, new ActionCompletionCallback (callback));
 		}
```
Comment (BrianBatchelder) [inline src/AirshipBindings.Android.Core/Additions/ActionRunRequest.cs]: SDK 10 API change.

File: src/AirshipBindings.Android.Core/Additions/RichPushInbox.cs [hunk 9]
```diff
@@ -38,7 +38,7 @@ public ICancelable FetchMessages(Action<bool> callback)
 
 		public ICancelable FetchMessages(Action<bool> callback, Looper looper)
 		{
-			return FetchMessages (new FetchMessagesCallback (callback), looper);
+			return FetchMessages (looper, new FetchMessagesCallback (callback));
```
Comment (BrianBatchelder) [inline src/AirshipBindings.Android.Core/Additions/RichPushInbox.cs]: SDK 10 API change.

File: src/AirshipBindings.Android.Core/Additions/UAirship.cs [hunk 10]
```diff
@@ -31,7 +31,7 @@ public static void TakeOff(Application application, AirshipConfigOptions configO
 
 		public static ICancelable Shared (Action<UAirship> callback, Looper looper)
 		{
-			return Shared (new AirshipReadyCallback (callback), looper);
+			return Shared (looper, new AirshipReadyCallback (callback));
```
Comment (BrianBatchelder) [inline src/AirshipBindings.Android.Core/Additions/UAirship.cs]: SDK 10 API change.

File: src/AirshipBindings.NETStandard/AirshipBindings.NETStandard.Android/Airship.cs [hunk 11]
```diff
@@ -143,7 +143,7 @@ public void AddCustomEvent(NETStandard.Analytics.CustomEvent customEvent)
                 }
             }
 
-            UAirship.Shared().Analytics.AddEvent(builder.Create());
+            UAirship.Shared().Analytics.AddEvent(builder.Build());
```
Comment (BrianBatchelder) [inline src/AirshipBindings.NETStandard/AirshipBindings.NETStandard.Android/Airship.cs]: SDK 10 API change.

File: src/AirshipBindings.Portable/AirshipBindings.Portable.Android/Airship.cs [hunk 12]
```diff
@@ -143,7 +143,7 @@ public void AddCustomEvent(Portable.Analytics.CustomEvent customEvent)
 				}
 			}
 
-			UAirship.Shared().Analytics.AddEvent(builder.Create());
+			UAirship.Shared().Analytics.AddEvent(builder.Build());
```
Comment (BrianBatchelder) [inline src/AirshipBindings.Portable/AirshipBindings.Portable.Android/Airship.cs]: SDK 10 API change.

File: samples/android/HomeFragment.cs [hunk 13]
```diff
@@ -71,28 +61,25 @@ public override void OnResume()
...
 			base.OnResume();
 
-			// Register a local broadcast manager to listen for ACTION_UPDATE_CHANNEL
-			LocalBroadcastManager localBroadcastManager = LocalBroadcastManager.GetInstance(Context);
-
-			// Use local broadcast manager to receive registration events to update the channel
-			IntentFilter channelUpdateFilter = new IntentFilter();
-			channelUpdateFilter.AddAction(UrbanAirshipReceiver.ACTION_CHANNEL_UPDATED);
-			localBroadcastManager.RegisterReceiver(channelIdUpdateReceiver, channelUpdateFilter);
+            UAirship.Shared().PushManager.AddRegistrationListener(this);
```
Comment (marc-scig) [inline samples/android/HomeFragment.cs]: This seems out of alignment

File: samples/android/HomeFragment.cs [hunk 14]
```diff
@@ -71,28 +61,25 @@ public override void OnResume()
...
-			string channelIdString = UAirship.Shared().PushManager.ChannelId;
+            UAirship.Shared().PushManager.RemoveRegistrationListener(this);
+        }
 
+        void RefreshChannelId()
+        {
+            RefreshChannelId(UAirship.Shared().PushManager.ChannelId);
+        }
+
+		void RefreshChannelId(String channelIdString)
```
Comment (marc-scig) [inline samples/android/HomeFragment.cs]: indentation is also weird here. maybe consider autoformatting?

File: samples/android/UrbanAirshipListener.cs [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+using Android.Support.V4.Content;
+
+namespace Sample
+{
+	public class UrbanAirshipListener : Java.Lang.Object, IRegistrationListener, IPushListener, INotificationListener
+    {
+		private const string TAG = "UrbanAirshipReceiver";
+
+        public void OnChannelCreated(String channelId)
+		{
```
Comment (marc-scig) [inline samples/android/UrbanAirshipListener.cs]: Same comment here regarding indentation/auto formatting

Comment (BrianBatchelder) [reply samples/android/UrbanAirshipListener.cs]: tabs...

File: samples/android/HomeFragment.cs (hunk 14 above)
Comment (BrianBatchelder) [reply samples/android/HomeFragment.cs]: tabs...

File: samples/android/HomeFragment.cs (hunk 13 above)
Comment (BrianBatchelder) [reply samples/android/HomeFragment.cs]: tabs..

**Comment by crow** [comment]
🍔

**Pull request merged**